### PR TITLE
Week mode

### LINKED ISSUE #39: Week view
Week view would be a nice feature to add.

Nice library btw,  thanks :)

### PR DESCRIPTION
**Pull request by andrey-ananiev**
I added a weekly mode to the calendar.
Closes #39

### CODE REVIEW

**Comment by boguszpawlowski** [comment]
Hi @andrey-ananiev. Thank you for your contribution! I'm a little short on time right now, but I should be able to review this on the weekend. Until then, please make sure that Detekt is passing and run `./gradlew apiDump` to generate new API scheme.

**Comment by andrey-ananiev** [comment]
Hi. I don`t know how I can retate the Detect Check. I am new to github.

File: library/src/main/java/io/github/boguszpawlowski/composecalendar/day/DayState.kt [hunk 1]
```diff
@@ -13,4 +14,5 @@ public typealias NonSelectableDayState = DayState<EmptySelectionState>
 public data class DayState<T : SelectionState>(
   private val day: Day,
   val selectionState: T,
+  val eventState: EventState
```
Comment (boguszpawlowski) [inline library/src/main/java/io/github/boguszpawlowski/composecalendar/day/DayState.kt:17]: I believe adding such functionality shouldn't be in scope of this PR.

File: library/src/main/java/io/github/boguszpawlowski/composecalendar/Calendar.kt [hunk 2]
```diff
@@ -75,11 +86,16 @@ public fun SelectableCalendar(
...
+  monthHeader: @Composable ColumnScope.(CurrentState) -> Unit = { DefaultMonthHeader(it) },
+  weekHeader: @Composable ColumnScope.(CurrentState) -> Unit = { DefaultWeekHeader(it) },
+  weekDaysNames: @Composable BoxScope.(List<DayOfWeek>) -> Unit = { DefaultWeekDaysNames(it) },
   monthContainer: @Composable (content: @Composable (PaddingValues) -> Unit) -> Unit = { content ->
     Box { content(PaddingValues()) }
   },
+  weekContainer: @Composable (content: @Composable (PaddingValues) -> Unit) -> Unit = { content ->
+    Box { content(PaddingValues()) }
+  },
+  onSwipe: (LocalDate) -> Unit = { },
```
Comment (boguszpawlowski) [inline library/src/main/java/io/github/boguszpawlowski/composecalendar/Calendar.kt:98]: Why is this necessary?

File: library/src/main/java/io/github/boguszpawlowski/composecalendar/header/DefaultMonthHeader.kt [hunk 3]
```diff
@@ -47,17 +48,17 @@ public fun DefaultMonthHeader(
     }
     Text(
       modifier = Modifier.testTag("MonthLabel"),
-      text = monthState.currentMonth.month
+      text = currentState.day.month
         .getDisplayName(FULL, Locale.getDefault())
         .lowercase()
         .replaceFirstChar { it.titlecase() },
-      style = MaterialTheme.typography.h4,
+      style = MaterialTheme.typography.h5,
```
Comment (boguszpawlowski) [inline library/src/main/java/io/github/boguszpawlowski/composecalendar/header/DefaultMonthHeader.kt:55]: What is the reason for this change?

File: library/src/main/java/io/github/boguszpawlowski/composecalendar/header/DefaultWeekDaysNames.kt [hunk 4]
```diff
@@ -1,4 +1,4 @@
-package io.github.boguszpawlowski.composecalendar.week
+package io.github.boguszpawlowski.composecalendar.header
```
Comment (boguszpawlowski) [inline library/src/main/java/io/github/boguszpawlowski/composecalendar/header/DefaultWeekDaysNames.kt:1]: Why changing the package?

File: library/src/main/java/io/github/boguszpawlowski/composecalendar/header/DefaultWeekDaysNames.kt [hunk 5]
```diff
@@ -11,12 +11,12 @@ import java.time.format.TextStyle.SHORT
 import java.util.Locale
 
 @Composable
-public fun DefaultWeekHeader(
-  daysOfWeek: List<DayOfWeek>,
+public fun DefaultWeekDaysNames(
+  listDays: List<DayOfWeek>,
```
Comment (boguszpawlowski) [inline library/src/main/java/io/github/boguszpawlowski/composecalendar/header/DefaultWeekDaysNames.kt:15]: Why changing the name?

File: library/src/main/java/io/github/boguszpawlowski/composecalendar/header/DefaultWeekHeader.kt [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+import androidx.compose.ui.platform.testTag
+import io.github.boguszpawlowski.composecalendar.states.CurrentState
+import java.time.format.TextStyle.FULL
+import java.util.Locale
+
+@Composable
+public fun DefaultWeekHeader(
+  currentState: CurrentState,
+  modifier: Modifier = Modifier,
+) {
```
Comment (boguszpawlowski) [inline library/src/main/java/io/github/boguszpawlowski/composecalendar/header/DefaultWeekHeader.kt:26]: Isn't this like a header for month / week depending on the situation? If this is the case, I think this name could be misleading.

File: library/src/main/java/io/github/boguszpawlowski/composecalendar/states/CurrentState.kt [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+import androidx.compose.runtime.Stable
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.saveable.Saver
+import androidx.compose.runtime.setValue
+import java.time.LocalDate
+
+@Suppress("FunctionName") // Factory function
+public fun CurrentState(initialDay: LocalDate): CurrentState = CurrentStateImpl(initialDay)
```
Comment (boguszpawlowski) [inline library/src/main/java/io/github/boguszpawlowski/composecalendar/states/CurrentState.kt:11]: I think `CurrentState` name is very vague. What does it represent? Is it a month / day / calendar state?

File: library/src/main/java/io/github/boguszpawlowski/composecalendar/month/MonthListState.kt [hunk 8]
```diff
@@ -5,33 +5,43 @@ import androidx.compose.runtime.Stable
...
   private val currentlyVisibleMonth by derivedStateOf {
     getMonthForPage(listState.firstVisibleItemIndex)
   }
 
   init {
-    snapshotFlow { monthState.currentMonth }.onEach { month ->
+    snapshotFlow { YearMonth.of(
+      currentState.day.year,
+      currentState.day.month
+    ) }.onEach { month ->
```
Comment (boguszpawlowski) [inline library/src/main/java/io/github/boguszpawlowski/composecalendar/month/MonthListState.kt:36]: Imo you should only pass the reference to the field you want to observe into `snapshotFlow`. If you want to map it later, you can employ `map` function of regular `Flow`

File: library/src/main/java/io/github/boguszpawlowski/composecalendar/month/MonthListState.kt [hunk 9]
```diff
@@ -5,33 +5,43 @@ import androidx.compose.runtime.Stable
...
       moveToMonth(month)
     }.launchIn(coroutineScope)
 
     snapshotFlow { currentlyVisibleMonth }.onEach { newMonth ->
-      monthState.currentMonth = newMonth
+      currentState.day = if (selectionState != null && selectionState.selection.isNotEmpty())
+        if (selectionState.selection[0].month == newMonth.month) newMonth.atDay(selectionState.selection[0].dayOfMonth)
+          else newMonth.atDay(1)
+      else newMonth.atDay(1)
```
Comment (boguszpawlowski) [inline library/src/main/java/io/github/boguszpawlowski/composecalendar/month/MonthListState.kt:44]: I think this logic is too complicated to be handled here. We should extract it to another place.

File: library/src/main/java/io/github/boguszpawlowski/composecalendar/states/CurrentState.kt (hunk 7 above)
Comment (boguszpawlowski) [inline library/src/main/java/io/github/boguszpawlowski/composecalendar/states/CurrentState.kt:11]: Also, I believe having the necessity to create such `CurrentState` is a side-effect of handling both `Month` and `Week` mode in the same composable. IMO this should be 2 separate composables - `Calendar` and `WeekCalendar`. This would allow us to avoid unnecessary complication of the code.

File: library/src/main/java/io/github/boguszpawlowski/composecalendar/states/EventState.kt [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+      )
+  }
+}
+
+
+
+
+
+
+//@Suppress("FunctionName") // Factory function
```
Comment (boguszpawlowski) [inline library/src/main/java/io/github/boguszpawlowski/composecalendar/states/EventState.kt:61]: Please remove commented code.

File: library/src/main/java/io/github/boguszpawlowski/composecalendar/week/WeekListState.kt [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+import io.github.boguszpawlowski.composecalendar.states.CurrentState
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.onEach
+import kotlinx.coroutines.launch
+import java.time.LocalDate
+import java.time.temporal.ChronoUnit
+
+@Stable
+internal class WeekListState(
```
Comment (boguszpawlowski) [inline library/src/main/java/io/github/boguszpawlowski/composecalendar/week/WeekListState.kt:17]: This and `WeekPager` could be easily used in order to create separate `WeekCalendar` composable.

File: library/src/main/java/io/github/boguszpawlowski/composecalendar/week/WeekUtil.kt [hunk 12]
```diff
@@ -48,3 +49,21 @@ internal fun YearMonth.getWeeks(
     )
   }
 }
+
+internal fun LocalDate.getWeek(
+  today: LocalDate = LocalDate.now(),
+): Week {
+  val firstDayOfWeek = minusDays(this.dayOfWeek.value.toLong() - 1)
```
Comment (boguszpawlowski) [inline library/src/main/java/io/github/boguszpawlowski/composecalendar/week/WeekUtil.kt:56]: Shouldn't this take into account the `firstDayOfTheWeek` parameter?

File: sample/src/main/java/io/github/boguszpawlowski/composecalendar/sample/CustomComponentsSample.kt [hunk 13]
```diff
@@ -24,23 +27,37 @@ import androidx.compose.ui.tooling.preview.Preview
...
+      modifier = Modifier.animateContentSize(),
+      showAdjacentMonths = false,
+      firstDayOfWeek = SUNDAY,
+      monthContainer = { MonthContainer(it) },
+      dayContent = { DayContent(dayState = it) },
+      weekDaysNames = { DefaultWeekDaysNames(listDays = it) },
+      monthHeader = { MonthHeader(currentState = it) },
+      calendarState = calendarState
+    )
+  }
```
Comment (boguszpawlowski) [inline sample/src/main/java/io/github/boguszpawlowski/composecalendar/sample/CustomComponentsSample.kt:60]: This cuould be in a separate `WeekCalendar` sample

File: sample/src/main/java/io/github/boguszpawlowski/composecalendar/sample/CustomSelectionSample.kt [hunk 14]
```diff
@@ -1,22 +1,42 @@
...
-  Calendar(calendarState = rememberMonthSelectionState())
+  val scrollState = rememberScrollState()
+  Column(modifier = Modifier
+    .fillMaxWidth()
+    .verticalScroll(scrollState)
+  ) {
+    val calendarState = rememberMonthSelectionState(
+      eventState = EventState( dayEventList )
+    )
+    ModeControls(modeState = calendarState.modeState)
```
Comment (boguszpawlowski) [inline sample/src/main/java/io/github/boguszpawlowski/composecalendar/sample/CustomSelectionSample.kt:37]: I don't think adding mode controls to every sample is necessary

**Review by boguszpawlowski — COMMENTED** [review]
Overall, I think it would be beneficial to change the approach a bit, and create a separate `WeekCalendar` composable. This would require a bit more of duplication but would allow us to simplify the logic which is pretty complicated right now. I think you could reuse `WeekPager` and `WeekListState` and after that it shouldn't be too much work.

**Comment by boguszpawlowski** [comment]
> Hi. I don`t know how I can retate the Detect Check. I am new to github
@andrey-ananiev If you have a mac, you can run `./gradlew detekt` (or `gradlew detekt`) locally in a terminal in project directory. This should show you all the errors from detekt.

File: library/src/main/java/io/github/boguszpawlowski/composecalendar/day/DayState.kt (hunk 1 above)
Comment (andrey-ananiev) [reply library/src/main/java/io/github/boguszpawlowski/composecalendar/day/DayState.kt:17]: In my project, I display the events of the day as dots in the calendar. Why not add this functionality to this PR

File: library/src/main/java/io/github/boguszpawlowski/composecalendar/Calendar.kt (hunk 2 above)
Comment (andrey-ananiev) [reply library/src/main/java/io/github/boguszpawlowski/composecalendar/Calendar.kt:98]: When the month changes, I load data to display dots representing events.

File: library/src/main/java/io/github/boguszpawlowski/composecalendar/header/DefaultMonthHeader.kt (hunk 3 above)
Comment (andrey-ananiev) [reply library/src/main/java/io/github/boguszpawlowski/composecalendar/header/DefaultMonthHeader.kt:55]: I think H4 is too big.

File: library/src/main/java/io/github/boguszpawlowski/composecalendar/states/EventState.kt (hunk 10 above)
Comment (andrey-ananiev) [reply library/src/main/java/io/github/boguszpawlowski/composecalendar/states/EventState.kt:61]: I'm sorry, I'll delete it.

File: library/src/main/java/io/github/boguszpawlowski/composecalendar/week/WeekListState.kt (hunk 11 above)
Comment (andrey-ananiev) [reply library/src/main/java/io/github/boguszpawlowski/composecalendar/week/WeekListState.kt:17]: Why do it separately when you can combine it?

File: library/src/main/java/io/github/boguszpawlowski/composecalendar/week/WeekUtil.kt (hunk 12 above)
Comment (andrey-ananiev) [reply library/src/main/java/io/github/boguszpawlowski/composecalendar/week/WeekUtil.kt:56]: I'm sorry... I'll fix it. I will definitely add this option.

File: sample/src/main/java/io/github/boguszpawlowski/composecalendar/sample/CustomComponentsSample.kt (hunk 13 above)
Comment (andrey-ananiev) [reply sample/src/main/java/io/github/boguszpawlowski/composecalendar/sample/CustomComponentsSample.kt:60]: I see that you are doing it in the next fork as you see fit. Tomorrow or the day after tomorrow I will have time to look at your varant in detail.

File: sample/src/main/java/io/github/boguszpawlowski/composecalendar/sample/CustomSelectionSample.kt (hunk 14 above)
Comment (andrey-ananiev) [reply sample/src/main/java/io/github/boguszpawlowski/composecalendar/sample/CustomSelectionSample.kt:37]: Fine.

File: library/src/main/java/io/github/boguszpawlowski/composecalendar/header/DefaultWeekDaysNames.kt (hunk 4 above)
Comment (andrey-ananiev) [reply library/src/main/java/io/github/boguszpawlowski/composecalendar/header/DefaultWeekDaysNames.kt:1]: I thought it made more sense.

File: library/src/main/java/io/github/boguszpawlowski/composecalendar/Calendar.kt [hunk 15]
```diff
@@ -172,45 +199,90 @@ public fun <T : SelectionState> Calendar(
...
-  val initialMonth = remember { calendarState.monthState.currentMonth }
+
   val daysOfWeek = remember(firstDayOfWeek) {
     DayOfWeek.values().rotateRight(DaysOfWeek - firstDayOfWeek.ordinal)
   }
 
   Column(
-    modifier = modifier,
+    modifier = modifier
+      .animateContentSize(),
```
[the remark this answers is not in the document]
Comment (andrey-ananiev) [reply library/src/main/java/io/github/boguszpawlowski/composecalendar/Calendar.kt:220]: Fine.

**Comment by boguszpawlowski** [comment]
@andrey-ananiev I have implemented something similar, using 2 separate composables, would be grateful if you would take a look and tell me what do you think about it: https://github.com/boguszpawlowski/ComposeCalendar/pull/91
Personally, I think this will be much easier to maintain in the future and the resulting code is much more readable.

File: library/src/main/java/io/github/boguszpawlowski/composecalendar/day/DayState.kt (hunk 1 above)
Comment (boguszpawlowski) [reply library/src/main/java/io/github/boguszpawlowski/composecalendar/day/DayState.kt:17]: Because this is easily achievable via the existing library API, and adding additional parameters to the API increases its complexity and reduces readability. Also, not every project will want to represent events as dots.

**Comment by andrey-ananiev** [comment]
> @andrey-ananiev I have implemented something similar, using 2 separate composables, would be grateful if you would take a look and tell me what do you think about it: #91 Personally, I think this will be much easier to maintain in the future and the resulting code is much more readable.

Thank you for supporting the weekly calendar idea. I will definitely look at your implementation and give feedback.

**Comment by boguszpawlowski** [comment]
@andrey-ananiev Thank you for your contribution, but I will be going with the second pr.
